### ISSUE TITLE
[Bug]: Last failed option does not work  with shards on single machine

### ISSUE THREAD

**Issue by romankhomitskyi** [microsoft/playwright#33804]
### Version

1.49.0

### Steps to reproduce

1. Clone my repo at https://github.com/romankhomitskyi/shard-failed
2. Run `npm ci`
3. Open `3` terminal windows
4. Execute the command `export SHARD_INDEX=1`, incrementing the index for each terminal (e.g, `export SHARD_INDEX=2`, `export SHARD_INDEX=3`)
5. Execute these commands in `3 `terminals accordingly
```
npx playwright test --project=chromium --shard=1/3
npx playwright test --project=chromium --shard=2/3
npx playwright test --project=chromium --shard=3/3
```
6. Some tests will fail
7. Execute these commands in `3 `terminals accordingly
```
npx playwright test --project=chromium --shard=1/3 --last-failed
npx playwright test --project=chromium --shard=2/3 --last-failed
npx playwright test --project=chromium --shard=3/3 --last-failed
```
You will see that only one test from one shard is run, rather than each from separate shard

### Expected behavior

I would like to run last failed test for each shard accordingly on single machine, locally

### Actual behavior

Even though I added this code to config
`outputDir: ./test-results-${process.env.SHARD_INDEX},`
To make sure shards do not override single test results, It does not run last failed tests for each generated test results

In short, it does not rerun only the failed tests, whether using a single test results folder or multiple separate ones, when using shards on single machine, locally

### Environment

```shell
System:
    OS: macOS 14.6.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 78.27 MB / 32.00 GB
  Binaries:
    Node: 20.15.1 - ~/.nvm/versions/node/v20.15.1/bin/node
    Yarn: 1.22.22 - ~/.nvm/versions/node/v20.15.1/bin/yarn
    npm: 10.7.0 - ~/.nvm/versions/node/v20.15.1/bin/npm
  IDEs:
    VSCode: 1.94.2 - /usr/local/bin/code
  Languages:
    Bash: 5.2.15 - /opt/homebrew/bin/bash
  npmPackages:
    @playwright/test: ^1.49.0 => 1.49.0
```

**Comment by dgozman** [comment]
Indeed, `--last-failed` filter is applied before the shard filter, so running with `--shard=1/3 --last-failed` runs one third of the last failed tests.

**Comment by dgozman** [comment]
@romankhomitskyi Thank you for the issue! Could you please explain your workflow that involves running `--last-failed` on multiple shards? We find this case very intriguing and wonder whether we can improve the story somehow? That said, we'll fix the exact `--last-failed --shard` issue anyway.

**Comment by romankhomitskyi** [comment]
> [@romankhomitskyi](https://github.com/romankhomitskyi) Thank you for the issue! Could you please explain your workflow that involves running `--last-failed` on multiple shards? We find this case very intriguing and wonder whether we can improve the story somehow? That said, we'll fix the exact `--last-failed --shard` issue anyway.

I would like to implement retry only failed tests, when retrying GitHub  job
I was testing it locally, basically reproducing the same workflow as it is on CI, and saw this issue, well, on CI it won't reproduce I guess, since each job is a separate machine

**Comment by romankhomitskyi** [comment]
Actually I managed to implement retry only failed tests when retrying CI job by uploading `.last-run.json`  as a artifact for each build and job, and then downloading it and checking in my `.sh` script

```
if [ -f "./test-results-$SHARD_INDEX/.last-run.json" ] && [ "$ATTEMPT" -gt 1 ]; then
  echo "Will run only failed specs"
  FAILED_TESTS_FLAG="--last-failed --pass-with-no-tests"
fi
```

`yarn playwright test --quiet --project=$PROJECTS --workers $WORKERS $FAILED_TESTS_FLAG`


 ```
 name: Download the failed specs if they are present 👀
        uses: actions/download-artifact@v4
        with:
          name: e2e_playwright_failed_specs_${{ github.run_id }}_${{ matrix.shardIndex }}
          path: ./qa/Playwright/test-results-${{ matrix.shardIndex }}
        continue-on-error: true



 name: Artifacts Playwright failed specs 👀
        uses: actions/upload-artifact@v4
        if: always()
        with:
          name: e2e_playwright_failed_specs_${{ github.run_id }}_${{ matrix.shardIndex }}
          retention-days: 1
          path: |
            ./qa/Playwright/test-results-${{ matrix.shardIndex }}/.last-run.json
          if-no-files-found: ignore
          include-hidden-files: true
```

**Comment by romankhomitskyi** [comment]
Currently, I do not use shards on CI or locally, I split specs into shards on my own, because I have special rules for where some specs should run, on the first shard or in the others

I just create multiple projects with filtered specs beforehand and that is the reason why `--last-failed `works for me locally

I just run `yarn playwright test --project=desktop-1 --last-failed` or `yarn playwright test --project=desktop-5 --last-failed`

The first shard should contain tests for feature flags, which mutate application state, so they run separately, and the other just randomly splitter randomly separated in the other shards (jobs) each job has its DB

**Comment by romankhomitskyi** [comment]
Then I was just testing the same workflow as mine with just shards and noticed it does not behave the same as my algorithm I though that some users might be testing locally with shards and won't get the same outcome

**Comment by dgozman** [comment]
Fixed by #34166.
